src/rust/ide/src/view/node_editor.rs

```diff
@@ -25,122 +92,293 @@ use weak_table::weak_value_hash_map::Entry::{Occupied, Vacant};
+            self.displayed_connections.borrow_mut().remove_rev(&edge_id);
         }
     }
+}
+
+
+// === Passing UI Actions To Controllers ===
+
+// These functions are called with FRP event values as arguments. The FRP values are always provided
+// by reference, even those "trivally-copy" types, To keep code cleaner we take all parameters
+// by reference as well.
+#[allow(clippy::trivially_copy_pass_by_ref)]
+impl GraphEditorIntegration {
+    fn node_removed_action(&self, node:&graph_editor::NodeId) -> FallibleResult<()> {
+        let id = self.get_controller_node_id(*node)?;
+        self.displayed_nodes.borrow_mut().remove_fwd(&id);
+        self.controller.graph.remove_node(id)?;
+        Ok(())
+    }
+
+    fn node_moved_action(&self, param:&(graph_editor::NodeId,frp::Position)) -> FallibleResult<()> {
+        let (displayed_id,pos) = param;
+        let id                 = self.get_controller_node_id(*displayed_id)?;
+        self.controller.graph.module.with_node_metadata(id, |md| {
+            md.position = Some(model::module::Position::new(pos.x,pos.y));
+        });
+        Ok(())
+    }
+
+    fn connection_created_action(&self, edge_id:&graph_editor::EdgeId) -> FallibleResult<()> {
+        let edge = self.editor.edges.get_cloned(&edge_id).ok_or(GraphEditorDiscrepancy)?;
+        let connection = self.from_graph_editor_edge(&edge)?;
+        self.displayed_connections.borrow_mut().insert(connection.clone(),*edge_id);
+        self.controller.graph.connect(&connection)?;
+        Ok(())
+    }
+
+    fn connection_removed_action(&self, edge_id:&graph_editor::EdgeId) -> FallibleResult<()> {
+        let connection = self.get_controller_connection(*edge_id)?;
+        self.displayed_connections.borrow_mut().remove_fwd(&connection);
+        self.controller.graph.disconnect(&connection)?;
+        Ok(())
+    }
+}
+
 
-    fn pos_to_vec3(pos:model::module::Position) -> Vector3<f32> {
-        Vector3::new(pos.vector.x,pos.vector.y,0.0)
+// === Utilities ===
+
+impl GraphEditorIntegration {
+    fn get_controller_node_id
+    (&self, displayed_id:graph_editor::NodeId) -> Result<ast::Id, MissingData> {
+        let err = MissingData::Node(displayed_id);
+        self.displayed_nodes.borrow().get_rev(&displayed_id).cloned().ok_or(err)
+    }
+
+    fn get_displayed_node_id
+    (&self, node_id:ast::Id) -> Result<graph_editor::NodeId, MissingData> {
+        let err = MissingData::DisplayedNode(node_id);
+        self.displayed_nodes.borrow().get_fwd(&node_id).cloned().ok_or(err)
+    }
+
+    fn get_controller_connection
+    (&self, displayed_id:graph_editor::EdgeId)
+    -> Result<controller::graph::Connection, MissingData> {
+        let err = MissingData::Connection(displayed_id);
+        self.displayed_connections.borrow().get_rev(&displayed_id).cloned().ok_or(err)
+    }
+
+    fn to_graph_editor_edge
+    (&self, connection:controller::graph::Connection) -> FallibleResult<(EdgeTarget,EdgeTarget)> {
```

Comment:
Shouldn't here (and in subsequent value usage) rather a struct for editor's edge be introduced — rather than having a tuple?

Reply:
This is the same type as GraphEditor is using.